### PR TITLE
Refs #24121 -- Added __repr__() to Engine.

### PR DESCRIPTION
**Pull request by abhiabhi94**

### CODE REVIEW

File: django/template/engine.py [hunk 1]
```diff
@@ -52,6 +52,18 @@ def __init__(self, dirs=None, app_dirs=False, context_processors=None,
         self.builtins = self.default_builtins + builtins
         self.template_builtins = self.get_template_builtins(self.builtins)
 
+    def __repr__(self):
+        return (
+            '<%(class)s dirs=%(dirs)s, app_dirs=%(app_dirs)s, loaders=%(loaders)s,'
+            ' context_processors=%(context_processors)s>'
+        ) % {
```
Comment (felixxm) [inline django/template/engine.py:59]: Please add all initialization parameters to the `__repr__()` and keep the order from `__init__()`. We should also use `repr()` for most attributes, e.g.
```python
    def __repr__(self):
        return (
            f'<{self.__class__.__qualname__} dirs={self.dirs!r} '
            f'app_dirs={self.app_dirs} '
            f'context_processors={self.context_processors!r} '
            f'debug={self.debug} loaders={self.loaders!r} '
            ...
        )
```

File: tests/template_tests/test_engine.py [hunk 2]
```diff
@@ -10,6 +10,36 @@
...
+            ' context_processors=%(context_processors)s>'
+        ) % {
+            'class': 'Engine',
+            'app_dirs': app_dirs,
+            'loaders': loaders,
+            'context_processors': context_processors,
+            'dirs': [TEMPLATE_DIR],
+        }
+        self.assertEqual(repr(engine), msg)
+
```
Comment (felixxm) [inline tests/template_tests/test_engine.py:41]: Add also a test for an "empty" engine i.e. `Engine()`.

File: tests/template_tests/test_engine.py [hunk 3]
```diff
@@ -10,6 +10,36 @@
...
+            '<%(class)s dirs=%(dirs)s, app_dirs=%(app_dirs)s, loaders=%(loaders)s,'
+            ' context_processors=%(context_processors)s>'
+        ) % {
+            'class': 'Engine',
+            'app_dirs': app_dirs,
+            'loaders': loaders,
+            'context_processors': context_processors,
+            'dirs': [TEMPLATE_DIR],
+        }
+        self.assertEqual(repr(engine), msg)
```
Comment (felixxm) [inline tests/template_tests/test_engine.py:40]: I would assert with an explicit representation (except `TEMPLATE_DIR` which depends on OS) e.g.
```python
        self.assertEqual(repr(engine), (
            f"<Engine dirs=[{TEMPLATE_DIR!r}] app_dirs=True "
            f"context_processors=[] loaders=["
            f"'django.template.loaders.filesystem.Loader' "
            f"'django.template.loaders.app_directories.Loader'] "
            ...
        ))
```

**Review by felixxm — COMMENTED** [review]
@abhiabhi94 Thanks :+1: I left comments.

File: tests/template_tests/test_engine.py (hunk 3 above)
Comment (abhiabhi94) [reply tests/template_tests/test_engine.py]: I have done what you asked for, but in my opinion this sort of long string with double quotes and single quotes mixed in between doesn't look too readable to me.

**Comment by abhiabhi94** [comment]
Thanks, @felixxm for the review. I have tried to address your concerns, please let me know if there's something more to it.

File: tests/template_tests/test_engine.py (hunk 3 above)
Comment (abhiabhi94) [reply tests/template_tests/test_engine.py]: Also, in this sort of assertions may be potentially hard to pin-point the accurate diff between long strings if the test fails. as it happened to me, I actually had to interchange the single quotes and double quotes for this test to pass i.e. when single quotes were put around the edges and double quotes for interpolation, the tests actually failed. I'm not really sure why this happened, but seemed quite weird to me.

Comment (abhiabhi94) [reply tests/template_tests/test_engine.py]: Just to reproduce what I said above, you can test this script:
```python
# t.py
import unittest

class LongStringTest(unittest.TestCase):
    a_list = ['a', 'b']
    a_tuple = ('a', 'b')
    a_complex = [('a', ['b', 'c'])]
    some_long_string = (
        f'List {a_list!r}'
        f'Tuple {a_tuple!r}'
        f'Complex {a_complex!r}'
    )
    def test_single_quotes_at_ends_with_double_quotes_for_interpolation(self):
        """This test fails"""
        self.assertEqual(
            self.some_long_string,
            (
                'List ["a", "b"]'
                'Tuple ("a", "b")'
                'Complex [("a", ["b", "c"])]'
            )
        )

    def test_double_quotes_at_ends_with_single_quotes_for_interpolation(self):
        """This test passes"""
        self.assertEqual(
            self.some_long_string,
            (
                "List ['a', 'b']"
                "Tuple ('a', 'b')"
                "Complex [('a', ['b', 'c'])]"
            )
        )
```
The output on my machine:
```sh
> python -m unittest t 
.F
======================================================================
FAIL: test_single_quotes_at_ends_with_double_quotes_for_interpolation (t.LongStringTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/django/tests/t.py", line 13, in test_single_quotes_at_ends_with_double_quotes_for_interpolation
    self.assertEqual(
AssertionError: "List ['a', 'b']Tuple ('a', 'b')Complex [('a', ['b', 'c'])]" != 'List ["a", "b"]Tuple ("a", "b")Complex [("a", ["b", "c"])]'
- List ['a', 'b']Tuple ('a', 'b')Complex [('a', ['b', 'c'])]
?       ^ ^  ^ ^        ^ ^  ^ ^           ^ ^   ^ ^  ^ ^
+ List ["a", "b"]Tuple ("a", "b")Complex [("a", ["b", "c"])]
?       ^ ^  ^ ^        ^ ^  ^ ^           ^ ^   ^ ^  ^ ^


----------------------------------------------------------------------
Ran 2 tests in 0.001s

FAILED (failures=1)

```

I think this is probably a bug in python maybe, will try to investigate further.

**Comment by felixxm** [comment]
@abhiabhi94 Thanks :+1:

I pushed small edits to omit empty attributes and make `repr()` more readable.

File: django/template/engine.py [hunk 4]
```diff
@@ -52,6 +52,26 @@ def __init__(self, dirs=None, app_dirs=False, context_processors=None,
...
+            self.app_dirs,
+            ''
+            if not self.context_processors
+            else ' context_processors=%s' % repr(self.context_processors),
+            self.debug,
+            repr(self.loaders),
+            repr(self.string_if_invalid),
+            repr(self.file_charset),
+            '' if not self.libraries else ' libraries=%s' % repr(self.libraries),
+            '' if not self.builtins else ' builtins=%s' % repr(self.builtins),
```
Comment (abhiabhi94) [inline django/template/engine.py:71]: I don't think `builtins` can be empty. the `default_builtins` should always be present?

**Comment by abhiabhi94** [comment]
Thanks.

File: django/template/engine.py (hunk 4 above)
Comment (felixxm) [reply django/template/engine.py:71]: They can be empty for subclasses, I think we can leave it that way.

Comment (abhiabhi94) [reply django/template/engine.py:71]: okay, then that branch needs a test too i guess?

Comment (felixxm) [reply django/template/engine.py:71]: It's only `__repr__()`, I don't think it's worth to test all cases.

**Pull request merged**